### ISSUE TITLE
[Feature Request] Strongly type require for `ColumnDef`

### ISSUE THREAD

**Issue by Goodjooy** [SeaQL/sea-query#315]
## Summary

When I am writting migration, I make a mistake following
```rust
ColumnDef::new(Id)
    .auto_increment()
    .primary_key()
    .not_null(),
```
I just forget giving the type of PK. But I do not capture this problem until runing migration. 
when constructing `ColumnDef`, having compile time limitation on column type assign

## Motivation

On define a column in a model. Enable this feature can prevent write the above code like me. It can alse prevent giving type mult times.
It is nearly not possible that a column without type 

## Additional Information

my Idea to impl this feature is using generic type help switch `types` from `()`(type not giving) to `ColumnType`(type given)

```rust
/// Specification of a table column
#[derive(Debug, Clone)]
pub struct ColumnDef<ColTy=ColumnType> {
    pub(crate) table: Option<TableRef>,
    pub(crate) name: DynIden,
    pub(crate) types: ColTy,
    pub(crate) spec: Vec<ColumnSpec>,
}

impl ColumnDef<()> {
    /// Construct a table column
    pub fn new<T: 'static>(name: T) -> Self
    where
        T: Iden,
    {
        ...
    }
}


impl ColumnDef {
    /// Construct a table column with column type
    pub fn new_with_type<T: 'static>(name: T, types: ColumnType) -> Self
    where
        T: Iden,
    {
        ...
    }
}

// impl for not giving type 
// need ownership, this might cause conflict
impl ColumnDef<()> {
    /// Set column type as char with custom length
    pub fn char_len(self, length: u32) -> ColumnDef {
        ...
    }
}

// impl for type given
impl ColumnDef {
    /// Set column unique constraint
    pub fn unique_key(&mut self) -> &mut Self {
        ...
    }
}
```

**Comment by billy1624** [comment]
Hey @Goodjooy, thanks for the proposal! I think we don't need generic, all we need is that we initialize `ColumnDef` with `ColumnType`. As you have already suggested...

```rs
impl ColumnDef {
    /// Construct a table column with column type
    pub fn new_with_type<T: 'static>(name: T, types: ColumnType) -> Self
    where
        T: Iden,
    {
        ...
    }
}
```

User can choose to opt-in to this new API for constructing new `ColumnDef`.

**Comment by Goodjooy** [comment]
> Hey @Goodjooy, thanks for the proposal! I think we don't need generic, all we need is that we initialize `ColumnDef` with `ColumnType`. As you have already suggested...
> 
> ```rust
> impl ColumnDef {
>     /// Construct a table column with column type
>     pub fn new_with_type<T: 'static>(name: T, types: ColumnType) -> Self
>     where
>         T: Iden,
>     {
>         ...
>     }
> }
> ```
> 
> User can choose to opt-in to this new API for constructing new `ColumnDef`.

Yes, I can use `new_with_type()` to create a `ColumnDef` with `ColumnType` 

But I 'd like to using help functions like `char_len()`, `string()` provided by `ColumnDef` to set column type. It will make `ColumnDef` in a partly constructing status when I have not set column type yet. 

It is necessary that having a strong limit on setting `ColumnType` of `ColumnDef` I believe. 

Thank you for your reply.
